test/utils/test_knowledge_utils.py

```diff
@@ -0,0 +1,357 @@
+    robot.pose = Pose.from_list([1.0, -0.5])
+    query = "kitchen table"
+    entity = query_to_entity(test_world, query.split(), "object", "nearest", robot)
+    assert (
+        entity.category == "banana"
+        and entity.parent.category == "table"
+        and entity.parent.parent.parent.name == "kitchen"
+    )
+
+    # we want the nearest apple on the kitchen table, should no longer return the banana even though it is the nearest object
+    query = "kitchen table apple"
+    new_entity = query_to_entity(test_world, query.split(), "object", "nearest", robot)
+    assert new_entity != entity
+
+    # things that should warn
+    import warnings
+
+    warnings.filterwarnings("error")
+    query = "kitchen table fake"
+    # search for nonexistent object in real location
+    for mode in ["object", "location"]:
+        try:
+            entity = query_to_entity(test_world, query.split(), mode)
+        except UserWarning:
+            # test passed, this should warn
+            pass
+        else:
+            assert False, "resolve_to_location didn't warn when not finding an object"
+
+    # search for absolute garbage
+    query = "fake fake fake"
+    for mode in ["object", "location"]:
+        try:
+            entity = query_to_entity(test_world, query.split(), mode)
+        except UserWarning:
+            # test passed, this should warn
+            pass
+        else:
+            assert False, "resolve_to_location didn't warn when not finding an object"
+
+
+def test_resolve_to_location():
+    test_world = load_world()
+    loc = resolve_to_location(test_world)
+    assert loc.name == "table0"
+
+    for category in ["table", "desk", "counter", "trash_can"]:
+        # we should be able to find each category
+        loc = resolve_to_location(test_world, category=category)
+        assert loc.category == category
+
+    for room in ["kitchen", "bedroom", "bathroom"]:
+        # all of these rooms have objects
+        loc = resolve_to_location(test_world, room=room)
+        assert loc.parent.name == room
+
+    # put ourselves on a table and make sure it is the nearest one
+    # then, search for a different category and make sure we find something that isn't the same table we're on
+    robot = Robot("test_robot")
+    robot.pose = Pose.from_list([0.85, -0.5, 0.0, -1.57])
+    nearest_loc = resolve_to_location(
+        test_world, resolution_strategy="nearest", robot=robot
+    )
+    assert nearest_loc.name == "table0"
+    loc = resolve_to_location(
+        test_world, category="desk", resolution_strategy="nearest", robot=robot
+    )
+    assert loc != nearest_loc
+
+    # test expanding locations
+    loc = resolve_to_location(
+        test_world,
+        category="desk",
+        resolution_strategy="nearest",
+        robot=robot,
+        expand_locations=True,
+    )
+    assert loc.parent.category == "desk"
+
+    # things that should warn
+    import warnings
+
+    warnings.filterwarnings("error")
```

Comment:
Again, if you import and set this up top you don't need to do it multiple times (within this test file, which seems OK to me)